jetson/ra_kinematics/test/kinematics_tests.cpp

```diff
@@ -0,0 +1,383 @@
+    for (size_t i = 0; i < 1; ++i) {
+        if (solver.IK(arm, target, false, false).second) {
+            success = true;
+            break;
+        } 
+    }
+
+    ASSERT_TRUE(success);
+}
+
+TEST(ik_local_min) {
+    json geom = read_json_from_file(get_mrover_arm_geom());
+    ArmState arm = ArmState(geom);
+    KinematicsSolver solver = KinematicsSolver();
+
+    arm.set_joint_angles({0, 1, -1, 0, 0, 0});
+    solver.FK(arm);
+    cout << "ef_pos after FK:\n";
+    cout << arm.get_ef_pos_world() << "\n";
+
+    // Create target point vector:
+    Vector6d target;
+    target << 38, 400, 14, 0.0, 0.0, 0.0;
+
+    solver.IK(arm, target, false, false);
+
+    cout << "num iterations: " << solver.get_num_iterations() << "\n";
+    ASSERT_FALSE(solver.get_num_iterations() > 500);
+}
+
+// Test that IK respects the locking of joints
+TEST(ik_test_lock) {
+    json geom = read_json_from_file(get_mrover_arm_geom());
+    ArmState arm = ArmState(geom);
+    KinematicsSolver solver = KinematicsSolver();
+
+    // Starting angles.
+    vector<double> start = {0, 1, -1, 0, 0, 0};
+
+    // Angles to find a target position.
+    vector<double> target = {0.1, 0.9, -1.1, 0.1, -0.1, 0};
+
+    ASSERT_TRUE(solver.is_safe(arm, start));
+    ASSERT_TRUE(solver.is_safe(arm, target));
+
+    // Generate the target position.
+    arm.set_joint_angles(target);
+    solver.FK(arm);
+
+    Vector6d target_pos;
+    target_pos.head(3) = arm.get_ef_pos_world();
+    target_pos.tail(3) = arm.get_ef_ang_world();
+
+    // Lock joint c
+    arm.set_joint_locked(2, true);
+
+    // Reset arm to starting position.
+    arm.set_joint_angles(start);
+    solver.FK(arm);
+
+    // Run IK
+    std::pair<Vector6d, bool> result = solver.IK(arm, target_pos, false, false);
+
+    // Check that joint c did not move from the start position.
+    ASSERT_ALMOST_EQUAL(-1, result.first[2], 0.0000001);
+
+    // Check that IK found a solution.
+    ASSERT_TRUE(result.second);
+}
+
+// Test that IK respects the locking of joints
+TEST(ik_test_orientation_experiment) {
+    json geom = read_json_from_file(get_mrover_arm_geom());
+    ArmState arm = ArmState(geom);
+    KinematicsSolver solver = KinematicsSolver();
+
+    // Starting angles.
+    vector<double> start = {0, 1, -1, 0, 0, 0};
+
+    // Angles to find a target position.
+    vector<double> target = {0.1, 0.9, -1, 0, 0, 0};
+
+    ASSERT_TRUE(solver.is_safe(arm, start));
+    ASSERT_TRUE(solver.is_safe(arm, target));
+
+    // Generate the target position.
+    arm.set_joint_angles(target);
+    solver.FK(arm);
+
+    Vector6d target_pos;
+    target_pos.head(3) = arm.get_ef_pos_world();
+    target_pos.tail(3) = arm.get_ef_ang_world();
+
+    // Reset arm to starting position.
+    arm.set_joint_angles(start);
+    solver.FK(arm);
+
+    // Run IK
+    std::pair<Vector6d, bool> result = solver.IK(arm, target_pos, false, true);
+
+    // Check that IK found a solution.
+    ASSERT_TRUE(result.second);
+}
```

Comment:
would it be possible to add tests for off nominal behavior to your testing suite? like simulating bad/inconsistent encoder values? or is that not easily done here

Reply:
That would not be easy. I definitely want to do this and we can probably figure out a way at some point, but we put a little thought into it sunday and couldn't think of a good solution right off the bat.